File: find-first-and-last-position-of-element-in-sorted-array_34.py
#TimeComplexity: logN + logN
#SpaceComplexity: O(1)
#It successfully runs on leetcode

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums or nums[0] > target or nums[-1] < target: return [-1,-1]
        low = 0
        high = len(nums)-1
        first=-1
        while low <= high:
            mid=low+(high-low)//2
            if nums[mid]==target:
                if mid==0 or (mid!=0 and nums[mid-1]<nums[mid]):
                    first = mid
                    break
                else:
                    high=mid-1
            elif nums[mid]<target:
                low=mid+1
            else:
                high=mid-1
        
        if first==-1: return [-1,-1]
                    
        low = first
        high = len(nums)-1
        while low<=high:
            mid=low+(high-low)//2
            if nums[mid]==target:
                if mid==len(nums)-1 or (mid!=len(nums)-1 and  nums[mid+1]>nums[mid]):
                    last=mid
                    break
                if nums[mid+1]==nums[mid]:
                    low=mid+1
            elif nums[mid]<=target:
                low=mid+1
            else:
                high=mid-1
        
        return [first,last]

File: Find First and Last Position of Element in Sorted Array-34.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        
        // If the nums array is null then we'll return [-1,-1] as it does not exist

        if(nums==null || nums.length==0)
        {
            return new int[]{-1,-1};
        }

// To check first occurence of target in num array
        int firstOccurence = firstOccurenceOfTarget(nums,target);
        // if there is no first occurence of the target in nums array then there wont be any second occurence so we return [-1,-1]
        if (firstOccurence==-1)
        {
            return new int[]{-1,-1};
        }
        // To check second occurence of target in num array
        int secondOcuurence = secondOccurenceOfTarget(firstOccurence, nums,target);
        return new int[]{firstOccurence, secondOcuurence};
    }


    private int firstOccurenceOfTarget(int[] nums, int target)
    {
        int low = 0;
        int high = nums.length-1;
        while(low<=high)
        {
            int mid = low+(high-low)/2; // Prevent Integer Overflow Exception
            if(nums[mid] == target) // Target found
            {
               if(mid == 0 || nums[mid-1]!=target) // If previous element from the mid is not target then we can return mid as first occurence 
               {
                   return mid; 
               }
               // If previous element from the mid is target then we continue with the process and check left side of nums array.
                high = mid-1;
            }
            else if (target>nums[mid]) // If target is greater than element at mid index then we move towards right
            {
                low = mid + 1;
            }
            else { // If target is less than element at mid index then we move towards left
                high = mid-1;
            }
        }
        return -1; // target not found
    }

    private int secondOccurenceOfTarget(int low, int[] nums, int target)
    {
    
      int high = nums.length-1;
      while(low<=high)
      {
          int mid = low+(high-low)/2;// Prevent Integer Overflow Exception
          if(nums[mid] == target)  // Target found
          {
              if( (mid==nums.length-1) || nums[mid+1]!=target )  // If next element from the mid is not target then we can return mid as last occurence 
              {
                  return mid;
              }
              // If next element from the mid is target then we continue with the process and check right side of nums array.
               low = mid +1; 
          }
          else if (target<nums[mid]) // If target is less than element at mid index then we move towards left
          {
             high = mid-1;
          }
          else { // If target is greater than element at mid index then we move towards right
           low = mid+1;
          }
      }
      return -1; // target not found
    }


}
